File: 11-Scheduling-And-Automation/EventOps/rule_engine.py
"""Rule Engine module for loading, validating, and matching EventOps rules."""

import json
import logging
import os
from dataclasses import dataclass

logger = logging.getLogger("eventops.rule_engine")


class RuleValidationError(Exception):
    """Custom exception raised when rule validation fails."""


@dataclass
class Rule:
    """Dataclass representing an EventOps automation rule."""

    name: str
    rule_type: str  # 'webhook' or 'file_change'
    action: str
    endpoint: str | None = None
    watch_path: str | None = None

    def __post_init__(self) -> None:
        if (
            self.rule_type == "webhook"
            and self.endpoint
            and not self.endpoint.startswith("/")
        ):
            self.endpoint = "/" + self.endpoint
        if self.rule_type == "file_change" and self.watch_path:
            self.watch_path = os.path.abspath(self.watch_path)


class RuleEngine:
    """Manages parsing, validating, and searching automation rules."""

    def __init__(self) -> None:
        self.rules: list[Rule] = []

    def load_from_file(self, file_path: str) -> list[Rule]:
        """Loads and validates rules from a JSON configuration file.

        Args:
            file_path: Path to rules.json.

        Returns:
            List of parsed Rule objects.

        Raises:
            RuleValidationError: If file does not exist or JSON schema is invalid.
        """
        if not os.path.exists(file_path):
            raise RuleValidationError(f"Configuration file not found: {file_path}")

        try:
            with open(file_path, "r", encoding="utf-8") as f:
                data = json.load(f)
        except json.JSONDecodeError as e:
            raise RuleValidationError(f"Invalid JSON format in {file_path}: {e}") from e
        except OSError as e:
            raise RuleValidationError(f"Failed to read {file_path}: {e}") from e

        return self.load_from_dict(data)

    def load_from_dict(self, data: dict) -> list[Rule]:
        """Loads and validates rules from a dictionary structure.

        Args:
            data: Parsed configuration dictionary.

        Returns:
            List of parsed Rule objects.

        Raises:
            RuleValidationError: If dict structure fails validation.
        """
        if not isinstance(data, dict) or "rules" not in data:
            raise RuleValidationError("Configuration root must contain a 'rules' list.")

        rules_list = data["rules"]
        if not isinstance(rules_list, list):
            raise RuleValidationError("'rules' must be a list of rule objects.")

        parsed_rules: list[Rule] = []
        rule_names = set()

        for idx, rule_data in enumerate(rules_list):
            if not isinstance(rule_data, dict):
                raise RuleValidationError(f"Rule at index {idx} must be a dictionary.")

            rule = self.validate_rule(rule_data, idx)
            if rule.name in rule_names:
                raise RuleValidationError(f"Duplicate rule name found: '{rule.name}'")
            rule_names.add(rule.name)
            parsed_rules.append(rule)

        self.rules = parsed_rules
        logger.info("Successfully loaded %d rules.", len(self.rules))
        return self.rules

    def validate_rule(self, rule_data: dict, idx: int) -> Rule:
        """Validates a single rule dictionary.

        Args:
            rule_data: Dictionary representing a rule.
            idx: Index of rule in configuration.

        Returns:
            Rule object.

        Raises:
            RuleValidationError: If required fields are missing or invalid.
        """
        name = rule_data.get("name")
        if not name or not isinstance(name, str):
            raise RuleValidationError(f"Rule at index {idx} missing valid 'name'.")

        rule_type = rule_data.get("type")
        if rule_type not in ("webhook", "file_change"):
            raise RuleValidationError(
                f"Rule '{name}' has invalid type '{rule_type}'. "
                "Must be 'webhook' or 'file_change'."
            )

        action = rule_data.get("action")
        if not action or not isinstance(action, str):
            raise RuleValidationError(f"Rule '{name}' missing valid 'action' command.")

        endpoint = None
        watch_path = None

        if rule_type == "webhook":
            endpoint = rule_data.get("endpoint")
            if not endpoint or not isinstance(endpoint, str):
                raise RuleValidationError(
                    f"Rule '{name}' of type 'webhook' requires a valid 'endpoint'."
                )
        elif rule_type == "file_change":
            watch_path = rule_data.get("watch_path")
            if not watch_path or not isinstance(watch_path, str):
                raise RuleValidationError(
                    f"Rule '{name}' of type 'file_change' requires a valid 'watch_path'."
                )

        return Rule(
            name=name,
            rule_type=rule_type,
            action=action,
            endpoint=endpoint,
            watch_path=watch_path,
        )

    def match_webhook(self, path: str) -> Rule | None:
        """Matches an incoming URI path against registered webhook rules.

        Args:
            path: Incoming HTTP URI path (e.g., '/deploy').

        Returns:
            Matching Rule if found, None otherwise.
        """
        normalized_path = path if path.startswith("/") else "/" + path
        # Normalize by removing trailing slash if not root
        if len(normalized_path) > 1 and normalized_path.endswith("/"):
            normalized_path = normalized_path.rstrip("/")

        for rule in self.rules:
            if rule.rule_type == "webhook" and rule.endpoint:
                rule_ep = rule.endpoint
                if len(rule_ep) > 1 and rule_ep.endswith("/"):
                    rule_ep = rule_ep.rstrip("/")
                if rule_ep == normalized_path:
                    return rule
        return None

    def match_file_change(self, modified_path: str) -> Rule | None:
        """Matches a modified file path against registered file_change rules.

        Args:
            modified_path: Path of the modified file.

        Returns:
            Matching Rule if found, None otherwise.
        """
        abs_modified = os.path.normcase(os.path.abspath(modified_path))

        for rule in self.rules:
            if (
                rule.rule_type == "file_change"
                and rule.watch_path
                and os.path.normcase(rule.watch_path) == abs_modified
            ):
                return rule
        return None

    def get_all_webhook_rules(self) -> list[Rule]:
        """Returns all rules of type 'webhook'."""
        return [r for r in self.rules if r.rule_type == "webhook"]

    def get_all_file_change_rules(self) -> list[Rule]:
        """Returns all rules of type 'file_change'."""
        return [r for r in self.rules if r.rule_type == "file_change"]

File: 11-Scheduling-And-Automation/EventOps-Go/pkg/ruleengine/ruleengine.go
// Package ruleengine parses, validates, and matches EventOps rules.
package ruleengine

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// Rule represents a single EventOps automation rule.
type Rule struct {
	Name      string `json:"name"`
	Type      string `json:"type"`
	Endpoint  string `json:"endpoint,omitempty"`
	WatchPath string `json:"watch_path,omitempty"`
	Action    string `json:"action"`
}

// Config represents the root structure of rules.json.
type Config struct {
	Rules []Rule `json:"rules"`
}

// RuleEngine handles rule storage, validation, and event matching.
type RuleEngine struct {
	rules []Rule
}

// NewRuleEngine initializes an empty RuleEngine.
func NewRuleEngine() *RuleEngine {
	return &RuleEngine{
		rules: make([]Rule, 0),
	}
}

// LoadFromFile reads and validates rules from a JSON file.
func (re *RuleEngine) LoadFromFile(filePath string) ([]Rule, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read config file %s: %w", filePath, err)
	}

	var cfg Config
	if err := json.Unmarshal(data, &cfg); err != nil {
		return nil, fmt.Errorf("invalid JSON schema in %s: %w", filePath, err)
	}

	return re.LoadFromConfig(&cfg)
}

// LoadFromConfig validates and registers rules from a Config struct.
func (re *RuleEngine) LoadFromConfig(cfg *Config) ([]Rule, error) {
	if cfg == nil || len(cfg.Rules) == 0 {
		return nil, fmt.Errorf("configuration rules list cannot be empty")
	}

	seenNames := make(map[string]bool)
	validated := make([]Rule, 0, len(cfg.Rules))

	for idx, r := range cfg.Rules {
		rule, err := re.validateRule(r, idx)
		if err != nil {
			return nil, err
		}

		if seenNames[rule.Name] {
			return nil, fmt.Errorf("duplicate rule name found: '%s'", rule.Name)
		}
		seenNames[rule.Name] = true
		validated = append(validated, rule)
	}

	re.rules = validated
	return re.rules, nil
}

func (re *RuleEngine) validateRule(r Rule, idx int) (Rule, error) {
	if strings.TrimSpace(r.Name) == "" {
		return r, fmt.Errorf("rule at index %d missing valid 'name'", idx)
	}

	if r.Type != "webhook" && r.Type != "file_change" {
		return r, fmt.Errorf("rule '%s' has invalid type '%s' (must be 'webhook' or 'file_change')", r.Name, r.Type)
	}

	if strings.TrimSpace(r.Action) == "" {
		return r, fmt.Errorf("rule '%s' missing valid 'action' command", r.Name)
	}

	if r.Type == "webhook" {
		if strings.TrimSpace(r.Endpoint) == "" {
			return r, fmt.Errorf("rule '%s' of type 'webhook' requires valid 'endpoint'", r.Name)
		}
		if !strings.HasPrefix(r.Endpoint, "/") {
			r.Endpoint = "/" + r.Endpoint
		}
	}

	if r.Type == "file_change" {
		if strings.TrimSpace(r.WatchPath) == "" {
			return r, fmt.Errorf("rule '%s' of type 'file_change' requires valid 'watch_path'", r.Name)
		}
		absPath, err := filepath.Abs(r.WatchPath)
		if err == nil {
			r.WatchPath = absPath
		}
	}

	return r, nil
}

// MatchWebhook searches for a rule matching an incoming URI path.
func (re *RuleEngine) MatchWebhook(path string) *Rule {
	normalized := path
	if !strings.HasPrefix(normalized, "/") {
		normalized = "/" + normalized
	}
	if len(normalized) > 1 && strings.HasSuffix(normalized, "/") {
		normalized = strings.TrimSuffix(normalized, "/")
	}

	for _, r := range re.rules {
		if r.Type == "webhook" && r.Endpoint != "" {
			ep := r.Endpoint
			if len(ep) > 1 && strings.HasSuffix(ep, "/") {
				ep = strings.TrimSuffix(ep, "/")
			}
			if ep == normalized {
				ruleCopy := r
				return &ruleCopy
			}
		}
	}
	return nil
}

// MatchFileChange searches for a rule matching a modified file path.
func (re *RuleEngine) MatchFileChange(modifiedPath string) *Rule {
	absMod, err := filepath.Abs(modifiedPath)
	if err != nil {
		absMod = modifiedPath
	}
	absModNorm := filepath.Clean(strings.ToLower(absMod))

	for _, r := range re.rules {
		if r.Type == "file_change" && r.WatchPath != "" {
			rulePathNorm := filepath.Clean(strings.ToLower(r.WatchPath))
			if rulePathNorm == absModNorm {
				ruleCopy := r
				return &ruleCopy
			}
		}
	}
	return nil
}

// GetWebhookRules returns all registered webhook rules.
func (re *RuleEngine) GetWebhookRules() []Rule {
	res := make([]Rule, 0)
	for _, r := range re.rules {
		if r.Type == "webhook" {
			res = append(res, r)
		}
	}
	return res
}

// GetFileChangeRules returns all registered file change rules.
func (re *RuleEngine) GetFileChangeRules() []Rule {
	res := make([]Rule, 0)
	for _, r := range re.rules {
		if r.Type == "file_change" {
			res = append(res, r)
		}
	}
	return res
}
